### PR TITLE
Export-Csv Test Improvements

### LINKED ISSUE #5139: Export-Csv Test Improvements
While looking at the current tests Export-Csv I found the following issues:

-----
https://github.com/PowerShell/PowerShell/blob/f604b0e3f063c7a89db025f11b320e5c9797c6e0/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1#L14
Should use `ShouldBeErrorId` and test for a proper FQEID

-----
https://github.com/PowerShell/PowerShell/blob/f604b0e3f063c7a89db025f11b320e5c9797c6e0/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1#L22
Should be `Should Be` and a full srting instead of partial match since this is static.

-----
https://github.com/PowerShell/PowerShell/blob/f604b0e3f063c7a89db025f11b320e5c9797c6e0/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1#L48
Should be `epcsv` here to test the alias.

-----
https://github.com/PowerShell/PowerShell/blob/f604b0e3f063c7a89db025f11b320e5c9797c6e0/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1#L69-L70
should be wrapped in a `BeforeAll{}`

-----

https://github.com/PowerShell/PowerShell/blob/f604b0e3f063c7a89db025f11b320e5c9797c6e0/test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1#L64
This can be removed because the `$TestDrive` will clean up after itself.

### PR DESCRIPTION
**Pull request by Marusyk**
Fix #5139

### CODE REVIEW

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 1]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
     $testCsv = Join-Path -Path $TestDrive -ChildPath "output.csv"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please put  the two lines in `BeforeAll` block.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 2]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
 
     AfterEach {
-	Remove-Item $testCsv -Force -ErrorAction SilentlyContinue
+        Remove-Item $testCsv -Force -ErrorAction SilentlyContinue
     }
 
     It "Should be able to be called without error" {
-	{ $testObject | Export-Csv $testCsv } | Should Not Throw
+        { $testObject | Export-Csv $testCsv } | Should Not Throw
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please add  `-ErrorAction Stop`.
```powershell
Export-Csv $testCsv -ErrorAction Stop
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 3]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
     }
 
     It "Should throw if an output file isn't specified" {
-	{ $testObject | Export-Csv -ErrorAction SilentlyContinue } | Should Throw
+        { $testObject | Export-Csv -ErrorAction SilentlyContinue } | ShouldBeErrorId "CannotSpecifyPathAndLiteralPath,Microsoft.PowerShell.Commands.ExportCsvCommand"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please use `-ErrorAction Stop`.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 4]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
 
     It "Should be a string when exporting via pipe" {
-	$testObject | Export-Csv $testCsv
+        $testObject | Export-Csv $testCsv
 
-	$piped = Get-Content $testCsv
+        $piped = Get-Content $testCsv
 
-	$piped[0] | Should Match ".String"
+        $piped[0] | Should be "#TYPE System.String"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please use:
```powershell
$piped[0] | Should BeExactly "#TYPE System.String" 
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 5]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
 
     It "Should be an object when exporting via the inputObject switch" {
-	Export-Csv -InputObject $testObject -Path $testCsv
+        Export-Csv -InputObject $testObject -Path $testCsv
 
-	$switch = Get-Content $testCsv
+        $switch = Get-Content $testCsv
 
-	$switch[0] | Should Match ".Object"
+        $switch[0] | Should Match ".Object"
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please use something like:
```powershell
$piped[0] | Should BeExactly "#TYPE System.Object" 
```

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 6]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
-	}
+        for ( $i = 0; $i -lt $testCsv.Length; $i++) {
+         $(Get-Content $testCsv)[$i] | Should Be $expected[$i]
+        }
     }
 
     It "Should be able to use the epcsv alias without error" {
-	{ $testObject | Export-Csv -Path $testCsv } | Should Not Throw
+        { $testObject | epcsv -Path $testCsv } | Should Not Throw
     }
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please remove the test at all - we have another file for alias tests.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 7]
```diff
@@ -3,71 +3,69 @@ Describe "Export-Csv" -Tags "CI" {
...
 
     It "Should have the same information when using the alias vs the cmdlet" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please remove the test at all - we have another file for alias tests.

Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 5 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 4 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 3 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 2 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 1 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 8]
```diff
@@ -1,73 +1,69 @@
...
+        $expected = @("#TYPE System.String", "`"Length`"", $first , $second, $third)
 
-    It "Should be able to use the epcsv alias without error" {
-	{ $testObject | Export-Csv -Path $testCsv } | Should Not Throw
+        for ( $i = 0; $i -lt $testCsv.Length; $i++) {
+         $(Get-Content $testCsv)[$i] | Should Be $expected[$i]
+        }
     }
 
     It "Should have the same information when using the alias vs the cmdlet" {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please remove the alias test.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 9]
```diff
@@ -1,73 +1,69 @@
...
-	}
-    }
+        $first    = "`"" + $testObject[0].Length.ToString() + "`""
+        $second   = "`"" + $testObject[1].Length.ToString() + "`""
+        $third    = "`"" + $testObject[2].Length.ToString() + "`""
+        $expected = @("#TYPE System.String", "`"Length`"", $first , $second, $third)
 
-    It "Should be able to use the epcsv alias without error" {
-	{ $testObject | Export-Csv -Path $testCsv } | Should Not Throw
+        for ( $i = 0; $i -lt $testCsv.Length; $i++) {
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: I wonder why we use `$testCsv.Length`?
I'd expect `$expected.Count`.
Your thoughts?

Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: I agree and think it should be $expected.Count. I have no clue how the length of the temporary file path string is relevant to anything in this test.

Comment (markekraus) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Is there even a need for looping logic here? would calling get-content once and perform 4 explicit tests be better? It seems like creating an array and looping through it (with a `for` loop no less) is just overkill.

Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: I think we should stop on ` $expected.Count`

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 (hunk 8 above)
Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1 [hunk 10]
```diff
@@ -1,73 +1,57 @@
...
-
-	for ( $i = 0; $i -lt $testCsv.Length; $i++)
-	{
-	    $(Get-Content $testCsv)[$i] | Should Be $(Get-Content $aliasObject)[$i]
-	}
-
-	# Clean up after yourself
-	Remove-Item $aliasObject -Force
+        for ( $i = 0; $i -lt $expected.Count; $i++) {
+         $(Get-Content $testCsv)[$i] | Should Be $expected[$i]
```
Comment (iSazonov) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Please correct indentation.

Comment (Marusyk) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Export-Csv.Tests.ps1]: Done

**Review by iSazonov — APPROVED** [review]

**Review by markekraus — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@Marusyk Please resolve merge conflicts.

**Comment by Marusyk** [comment]
@iSazonov Done

**Comment by markekraus** [comment]
appveyor build fail due to #5201

**Comment by Marusyk** [comment]
@markekraus But error about version of .NET Command Line Tools. It doesn't depends on this changes

**Comment by markekraus** [comment]
@Marusyk Right. that's what I said. the AppVeyor failure is due to the issue noted in #5201

**Comment by daxian-dbw** [comment]
#5118 has been merged, and I restarted AppVeyor CI for this PR.

**Review by markekraus — APPROVED** [review]
LGTM

**Comment by iSazonov** [comment]
@daxian-dbw Can we merge the PR?

**Comment by daxian-dbw** [comment]
@iSazonov Sorry that I didn't get a chance to review this change (too many notifications since having the code owner filter ...). Please proceed as you see appropriate :)

**Comment by iSazonov** [comment]
@Marusyk Thanks for your contribution! Welcome back with new PRs.

**Pull request merged**